File: backend/src/modules/location/location.controller.ts
import {
  Controller,
  Get,
  Delete,
  Query,
  Param,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse } from '@nestjs/swagger';
import { LocationService } from './location.service';
import { PaginationDto } from '../../common/dto/pagination.dto';
import { ApiResponseDto } from '../../common/dto/api-response.dto';

@ApiTags('注册位置管理')
@Controller('api/locations')
export class LocationController {
  constructor(private readonly locationService: LocationService) {}

  @Get()
  @ApiOperation({ summary: '获取注册位置列表' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async findAll(@Query() paginationDto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.locationService.findAll(paginationDto);
    return ApiResponseDto.success(result);
  }

  @Get('memory')
  @ApiOperation({ summary: '从内存获取注册位置' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getFromMemory(@Query('table') table?: string): Promise<ApiResponseDto> {
    const result = await this.locationService.getFromMemory(table);
    return ApiResponseDto.success(result);
  }

  @Get('stats')
  @ApiOperation({ summary: '获取注册统计信息' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getStats(): Promise<ApiResponseDto> {
    const stats = await this.locationService.getStats();
    return ApiResponseDto.success(stats);
  }

  @Get('online-count')
  @ApiOperation({ summary: '获取在线用户数量' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getOnlineCount(): Promise<ApiResponseDto<number>> {
    const count = await this.locationService.getOnlineCount();
    return ApiResponseDto.success(count);
  }

  @Get('user/:username')
  @ApiOperation({ summary: '根据用户名查询注册' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async findByUsername(@Param('username') username: string): Promise<ApiResponseDto> {
    const locations = await this.locationService.findByUsername(username);
    return ApiResponseDto.success(locations);
  }

  @Delete(':username/:domain')
  @ApiOperation({ summary: '删除用户注册' })
  @ApiResponse({ status: 200, description: '删除成功' })
  async deleteUserLocation(
    @Param('username') username: string,
    @Param('domain') domain: string,
  ): Promise<ApiResponseDto> {
    await this.locationService.deleteUserLocation(username, domain);
    return ApiResponseDto.success(null, '删除成功');
  }

  @Get('lookup')
  @ApiOperation({ summary: '精确查找指定用户注册 (RPC ul.lookup)' })
  async lookupUser(
    @Query('table') table: string,
    @Query('aor') aor: string,
  ): Promise<ApiResponseDto> {
    return ApiResponseDto.success(await this.locationService.lookupUser(table || 'location', aor));
  }

  @Delete('contact')
  @ApiOperation({ summary: '删除指定联系地址 (RPC ul.rm_contact)' })
  async deleteContact(
    @Body() body: { table: string; aor: string; contact: string },
  ): Promise<ApiResponseDto> {
    await this.locationService.deleteContact(body.table || 'location', body.aor, body.contact);
    return ApiResponseDto.success(null, '删除成功');
  }
}

File: kam-admin-module-session/src/main/java/io/kamailio/admin/modules/location/controller/LocationController.java
package io.kamailio.admin.modules.location.controller;

import io.kamailio.admin.common.dto.ApiResponse;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.location.entity.Location;
import io.kamailio.admin.modules.location.service.LocationService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@Tag(name = "注册位置管理")
@RestController
@RequestMapping("/api/locations")
public class LocationController {
    private final LocationService service;

    public LocationController(LocationService service) { this.service = service; }

    @GetMapping
    @Operation(summary = "获取注册位置列表")
    public ResponseEntity<ApiResponse<PaginatedResult<Location>>> findAll(PaginationDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.findAll(dto)));
    }

    @GetMapping("/memory")
    @Operation(summary = "从内存获取注册位置")
    public ResponseEntity<ApiResponse<Object>> getFromMemory(@RequestParam(required = false) String table) {
        return ResponseEntity.ok(ApiResponse.success(service.getFromMemory(table)));
    }

    @GetMapping("/stats")
    @Operation(summary = "获取注册统计信息")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getStats() {
        return ResponseEntity.ok(ApiResponse.success(service.getStats()));
    }

    @GetMapping("/online-count")
    @Operation(summary = "获取在线用户数量")
    public ResponseEntity<ApiResponse<Long>> getOnlineCount() {
        return ResponseEntity.ok(ApiResponse.success(service.getOnlineCount()));
    }

    @GetMapping("/user/{username}")
    @Operation(summary = "根据用户名查询注册")
    public ResponseEntity<ApiResponse<List<Location>>> findByUsername(@PathVariable String username) {
        return ResponseEntity.ok(ApiResponse.success(service.findByUsername(username)));
    }

    @DeleteMapping("/{username}/{domain}")
    @Operation(summary = "删除用户注册")
    public ResponseEntity<ApiResponse<Void>> deleteUserLocation(@PathVariable String username, @PathVariable String domain) {
        service.deleteUserLocation(username, domain);
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @GetMapping("/lookup")
    @Operation(summary = "精确查找指定用户注册")
    public ResponseEntity<ApiResponse<Object>> lookupUser(@RequestParam(required = false) String table, @RequestParam String aor) {
        return ResponseEntity.ok(ApiResponse.success(service.lookupUser(table, aor)));
    }

    @DeleteMapping("/contact")
    @Operation(summary = "删除指定联系地址")
    public ResponseEntity<ApiResponse<Void>> deleteContact(@RequestBody Map<String, String> body) {
        service.deleteContact(body.get("table"), body.get("aor"), body.get("contact"));
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @PostMapping("/flush")
    @Operation(summary = "将 usrloc 内存缓存刷入数据库")
    public ResponseEntity<ApiResponse<Void>> flushUsrloc() {
        service.flushUsrloc();
        return ResponseEntity.ok(ApiResponse.success(null, "刷新成功"));
    }
}
